### PR TITLE
Issue #3173 fix.

### PR DESCRIPTION
**Pull request by shmuely**
(#3173) Context menu now passes hovered nodes & edges.

### CODE REVIEW

File: lib/network/modules/InteractionHandler.js [hunk 1]
```diff
@@ -165,6 +165,7 @@ class InteractionHandler {
 
   onContext(event) {
     let pointer = this.getPointer({x:event.clientX, y:event.clientY});
+    this.checkSelectionChanges(pointer, event);
```
Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: What you're doing here is setting the selection to the node currently hovered over.
This is not what you want; you just want to know over which node the pointer is hovering.

I propose that you use `Selectionhandler.getNodeAt()` instead.

Comment (shmuely) [inline lib/network/modules/InteractionHandler.js]: Got you, I changed it to `Selectionhandler.getNodeAt()` method.

File: lib/network/modules/InteractionHandler.js [hunk 2]
```diff
@@ -165,8 +165,11 @@ class InteractionHandler {
 
   onContext(event) {
     let pointer = this.getPointer({x:event.clientX, y:event.clientY});
-    this.checkSelectionChanges(pointer, event);
-    this.selectionHandler._generateClickEvent('oncontext', event, pointer);
+    const hoveredElements = {
+      nodes: this.selectionHandler.getNodeAt(pointer.DOM),
+      edges: this.selectionHandler.getEdgeAt(pointer.DOM)
```
Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: There is no DOM object within pointer, just x and y. Use the pointer object directly.

`nodes` and `edges` are arrays, please add any hovered over nodes and edges as array, and *only if these are present*; i.e. `getNodeAt()` and `getEdgeAt()` may return `undefined`.

File: lib/network/modules/InteractionHandler.js [hunk 3]
```diff
@@ -165,8 +165,11 @@ class InteractionHandler {
 
   onContext(event) {
```
Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: Event `oncontext` triggers on a right-click; your issue seems to suggest that you want to determine the nodes and edges that are hovered over. In that case, events `hoverNode` and `hoverEdge` are more appropriate; these already return the node and edge id's in question.

Please confirm it's really 'oncontext' you need.

File: lib/network/modules/InteractionHandler.js [hunk 4]
```diff
@@ -165,8 +165,11 @@ class InteractionHandler {
...
     let pointer = this.getPointer({x:event.clientX, y:event.clientY});
-    this.checkSelectionChanges(pointer, event);
-    this.selectionHandler._generateClickEvent('oncontext', event, pointer);
+    const hoveredElements = {
+      nodes: this.selectionHandler.getNodeAt(pointer.DOM),
+      edges: this.selectionHandler.getEdgeAt(pointer.DOM)
+    }
+    this.selectionHandler._generateClickEvent('oncontext', Object.assign({}, event, hoveredElements), pointer);
```
Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: Personal note: that's a pretty nifty use of `Object.assign()`. Wouldn't have thought of it myself, TIL.

File: lib/network/modules/InteractionHandler.js (hunk 3 above)
Comment (shmuely) [inline lib/network/modules/InteractionHandler.js]: Yes, it's what I need, i'll explain, when user clicks right-click, he does not have access to `hoverNode`'s params, so on right click I would like to access the node/edge that i'm hovering.
:)

Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: OK clear. For completeness, I mention that `getNodeAt()` would have worked in the callback handler, as the documentation says.

File: lib/network/modules/InteractionHandler.js [hunk 5]
```diff
@@ -165,8 +165,13 @@ class InteractionHandler {
 
   onContext(event) {
     let pointer = this.getPointer({x:event.clientX, y:event.clientY});
-    this.checkSelectionChanges(pointer, event);
-    this.selectionHandler._generateClickEvent('oncontext', event, pointer);
+    const hoveredElements = {
+      nodes: (this.selectionHandler.getNodeAt(pointer) || []),
```
Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: You are using the entire node and edge object here. What should be set just just the node and edge id's (conforming to documentation).

File: lib/network/modules/InteractionHandler.js [hunk 6]
```diff
@@ -165,8 +165,13 @@ class InteractionHandler {
...
+      edges: (this.selectionHandler.getEdgeAt(pointer) || [])
+    }
+    hoveredElements.nodes = [].concat(hoveredElements.nodes)
+    hoveredElements.edges = [].concat(hoveredElements.edges)
+    this.selectionHandler._generateClickEvent('oncontext', Object.assign({}, event, hoveredElements), pointer);
```
Comment (wimrijnders) [inline lib/network/modules/InteractionHandler.js]: One I should have spotted earlier: `nodes` and `edges` are defined here as part of the event. In the output, they are not. `_generateClickEvent()` create the `properties` object, adds `nodes` and `edges` to it, and the original event (which you are changing) is made a *member* of this `properties` object.

**Comment by wimrijnders** [comment]
@shmuely While I admire your js-skills, I think this PR is going nowhere.

I strongly urge you to consider calling `getNodeAt()` and `getEdgeAt()` in the event handler for `oncontext`. It's a much simpler and cleaner solution.
I you wish to persevere, I will keep on reviewing, but I have no intentions of relaxing the quality standards.

On a personal note, I will add that I am enormously happy that you are taking the trouble of changing the code yourself. This seldom happens. I am the only regular coder for `network` and sometimes I do wonder how all the submitters think how I can address all their issues by myself. So really thank you for that.

**Comment by yotamberk** [comment]
+1 :point_up:

**Comment by wimrijnders** [comment]
@shmuely Are you going to keep on working on this PR? If not, please close it.

Also, if you stop, I'm sorry if I have discouraged you. But please realize it is my duty as a reviewer to keep a quality standard. I then wish you a better success next time.

**Comment by shmuely** [comment]
Sorry for the delay, I moved the functionality to InteractionHandler.js

**Comment by wimrijnders** [comment]
@shmuely Can you fix the conflicts, pleas?

**Comment by shmuely** [comment]
@wimrijnders - conflicts resolved

**Comment by yotamberk** [comment]
@wimrijnders Can you please take this?

**Comment by wimrijnders** [comment]
@yotamberk yes when I get back from vacation. That goes for the rest as well.

**Comment by wimrijnders** [comment]
@yotamberk yes when I get back from vacation. That goes for the rest as well.

**Comment by yotamberk** [comment]
No rush! ENJOY your vacation!

**Comment by wimrijnders** [comment]
@shmuely Apologies for the delay, my mind was on other things and there was also a vacation in between.

**But:** The code is not working as intended. Please test first before submitting to ensure proper working, and then I'll review. It might be an idea to create a unit test for this.

**Comment by macleodbroad-wf** [comment]
Given the lack of movement here, I'd prefer this be closed and re-opened when it is ready or requires feedback.

**Comment by wimrijnders** [comment]
Closing; aside from non-activity, the code has progressed to the point that this fix is irrelevant.

**Pull request closed without merging**